### ISSUE TITLE
Add kafka serializer using confluent schema registry

### ISSUE THREAD

**Issue by Chuckame** [avro-kotlin/avro4k#218]
Currently, to be able to use the kafka-avro-serializer from confluent (using a schema registry under the hood), we need to use `Avro.encodeToGenericData` + `KafkaAvroSerializer.serialize`, and `KafkaAvroDeserializer.deserialize` + `Avro.decodeFromGenericData`.

There are 2 main concerns here:
- The performance is pretty bad, as we need to first ~encode~ convert the serializable data to generic data, then serializing the generic data to avro bytes.
- There is basically 2 times more code to maintain in avro4k, and this generic data feature is not tested anymore, prior to the binary avro encoding.

This confluent serializer is not doing a big work: it ensures having the good schema and adapt if needed the reader schema to the written data's schema, and it prepend the schema id to the avro bytes, that's all.

Avro4k is already able to adapt on-the-fly the decoded data's schema, and also able to adapt the encoding depending on the target writer schema. So the only remaining work is to handle basic features of the serializer on the schema registry (auto register, use specific schema id, use latest registered schema) and prepend the retrieved schema id before the avro bytes.

The main issue is that there is no way to plug avro4k to the confluent's serializer (even doing tricks with `SpecificRecord`), and their method to find/register the schema and prepend the schema id is private.

So we have to do it !

--- 

We will have a new module with 3 new serializers:
- `Avro4kKafkaSerde` (and the corresponding serializer/deserializer): a serializer based on a specific and unique serializable type. Helpful for apps like kafka-streams or simple spring kafka consumer/producer, knowing at compile-time the type of a topic key or value by example
- `GenericAvro4kKafkaSerde` (and the corresponding serializer/deserializer): a serializer that can serialize and deserialize any type supported by kotlinx-serialization. This is helpful when not having data classes or enums known by advance, allowing generic consumption and production of events. This is plugged to the confluent `GenericContainer` classes carrying the schema.
- `ReflectAvro4kKafkaSerde` (and the corresponding serializer/deserializer): nearly the same features as the generic one, but by using concrete serializable data classes and enums using reflection. Warning: you would need the concrete classes, or deserialization would fail!

**Comment by jamesforward999** [comment]
Hey @Chuckame, do you have any details on when this will be made available?

**Comment by Chuckame** [comment]
@jamesforward999 there isn't any ETA for the moment as I'm working on that project during my free time. I would like to release a first beta version for the next month. The main time consuming work here is to have a great test coverage and a consistent API.

**Comment by SergiioDev** [comment]
Hey @Chuckame this will work with Apicurio or only confluence schema registry?

**Comment by Chuckame** [comment]
@SergiioDev honestly I suppose avro4k is not a good player with apicurio, as it seems the data is serialized in json and not in avro binary format. What's the link between avro and apicurio?

**Comment by SergiioDev** [comment]
So I use apicurio registry to register our avro schemas and I was wondering if this feature will work with that registry too

**Comment by Chuckame** [comment]
No, because the confluent serializer is prefixing some bytes to the real avro payload with the schema id, and is plugged to their rest API to register and retrieve schemas. To plug apicurio, you need to comply to the confluent API... So maybe using the confluent schema registry instead would be simpler I suppose

**Comment by Chuckame** [comment]
After deeper reading, it seems that apicurio is 100% compatible with confluent SR api, and also serializes the same way as confluent with a magic byte + a 4 bytes schema id ([link](https://github.com/Apicurio/apicurio-registry/blob/main/serdes/generic/serde-common/src/main/java/io/apicurio/registry/serde/AbstractSerializer.java#L20)). So the module I'm creating may be natively compatible with apicurio. When the module will be released, don't hesitate to give me feedback!

**Comment by SergiioDev** [comment]
Amazing, thank you for the response I'll try this and give you feedback!!

**Comment by Chuckame** [comment]
Hello there, the new module has been created, but I'm struggling with the release...

If anyone of

**Comment by Chuckame** [comment]
Hello there, after nearly a year of thinking and months of work (open source and personal life takes time sometimes 😄), the new module has been created!! 🚀 

But I'm struggling with the release... 😅 

If anyone of you want to give it a try by compiling the sources locally, tell me ! Until that time, I'm doing my max to make it work again (credentials issue)

**Comment by Chuckame** [comment]
Have been released in v2.4.0
